### PR TITLE
Moving from PHP serialized variables to JSON encoded variables

### PR DESCRIPTION
**Pull request by Skyld**
Storing outcome, response and trace variables as JSON strings instead of PHP serialized objects in the database is the first step in the process of having an ETL solution on top of TAO for easy data extraction.

This implementation is meant to be a backward compatible change, meaning that if the attempt of json decoding the raw variable fails, it will fall back to the original object unserialization.

Related PR: https://github.com/oat-sa/extension-tao-outcomerds/pull/110

### CODE REVIEW

File: models/classes/class.Variable.php [hunk 1]
```diff
@@ -18,182 +21,235 @@
...
+    }
+
+    /**
+     * {@inheritdoc}
      */
-    public function toJson()
+    public function jsonSerialize()
     {
-        return json_encode((array)$this);
+        return [
```
Comment (luisfn) [inline models/classes/class.Variable.php]: you could use return `get_object_vars($this)` probably since you seems to return all properties

File: models/classes/class.Variable.php [hunk 2]
```diff
@@ -18,182 +21,235 @@
...
+     */
+    public static function fromData(array $rawVariable)
+    {
+        if (!isset($rawVariable['type'])) {
+            throw new \LogicException('Variable type declaration is missing.');
+        }
+
+        self::validateKeys(['identifier', 'cardinality', 'baseType', 'epoch'], $rawVariable);
+
+        switch ($rawVariable['type']) {
```
Comment (luisfn) [inline models/classes/class.Variable.php]: Maybe we could have a factory class for returning these?

File: models/classes/class.Variable.php [hunk 3]
```diff
@@ -18,182 +21,235 @@
...
+            'epoch' => $this->epoch,
+        ];
+    }
+
+    /**
+     * @return taoResultServer_models_classes_OutcomeVariable|taoResultServer_models_classes_ResponseVariable|taoResultServer_models_classes_TraceVariable
+     * @throws common_exception_InvalidArgumentType
+     * @throws LogicException
+     */
+    public static function fromData(array $rawVariable)
```
Comment (andreluizmachado) [inline models/classes/class.Variable.php]: I would encourage you to extract all these factory methods to a new class that just construct Variable objects. Like it is, the abstraction knows about the concrete objects and also because we could remove one responsibility from this class.

File: manifest.php [hunk 4]
```diff
@@ -13,7 +13,7 @@
     'label' => 'Result core extension',
     'description' => 'Results Server management and exposed interfaces for results data submission',
     'license' => 'GPL-2.0',
-    'version' => '10.4.0',
+    'version' => '10.5.0',
```
[the remark this answers is not in the document]
Comment (Skyld) [reply manifest.php]: I tried to keep backward compatibility, do you see any breaking change?

[the remark this answers is not in the document]
Comment (Skyld) [reply manifest.php]: Good point

File: models/classes/class.Variable.php [hunk 5]
```diff
@@ -18,182 +21,238 @@
...
     abstract public function setValue($value);
 
     /**
-     * Get the json representation of the variable
-     * @return string
+     * @deprecated Use jsonSerialize method instead
+     */
+    public function toJson(): string
+    {
+        return json_encode($this->jsonSerialize());
```
Comment (jbout) [inline models/classes/class.Variable.php]: might be clearer to just `return json_encode($this)` as this wil ltrigger the jsonSerialize

File: test/Unit/Models/classes/class.OutcomeVariableTest.php [hunk 6]
```diff
@@ -0,0 +1,51 @@
+<?php
+
+declare(strict_types=1);
+
+namespace oat\taoResultServer\test\Unit\Models\classes;
```
Comment (jbout) [inline test/Unit/Models/classes/class.OutcomeVariableTest.php]: Why did you rename `models` to `Models`? Now it no longer matches the tested classes

File: manifest.php (hunk 4 above)
[the remark this answers is not in the document]
Comment (jbout) [reply manifest.php]: Have we validated if this will not break some of our customers code?

File: models/classes/class.Variable.php (hunk 5 above)
Comment (Skyld) [reply models/classes/class.Variable.php]: Ah, indeed, good catch

File: models/classes/class.Variable.php [hunk 7]
```diff
@@ -18,182 +21,235 @@
...
+    }
+
+    /**
+     * @return taoResultServer_models_classes_OutcomeVariable|taoResultServer_models_classes_ResponseVariable|taoResultServer_models_classes_TraceVariable
+     * @throws common_exception_InvalidArgumentType
+     * @throws LogicException
+     */
+    public static function fromData(array $rawVariable)
+    {
+        if (!isset($rawVariable['type'])) {
```
[the remark this answers is not in the document]
Comment (Skyld) [reply models/classes/class.Variable.php]: Fixed

File: test/Unit/Models/classes/class.OutcomeVariableTest.php (hunk 6 above)
Comment (Skyld) [reply test/Unit/Models/classes/class.OutcomeVariableTest.php]: Reverted
